File: Example/FWFramework/Classes/Module/Test/Service/TestCacheController.m
//
//  TestCacheController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestCacheController.h"
#import "AppSwift.h"
@import FWFramework;

#define TestCacheKey @"TestCacheKey"
#define TestExpireKey @"TestCacheKey.__EXPIRE__"

@interface TestCacheController () <FWViewController>

@property (nonatomic, strong) id<FWCacheProtocol> cache;

@property (nonatomic, strong) UILabel *cacheLabel;

@end

@implementation TestCacheController

- (void)setupNavbar
{
    FWWeakifySelf();
    [self fw_setRightBarItem:@"切换" block:^(id sender) {
        FWStrongifySelf();
        
        [self fw_showSheetWithTitle:@"选择缓存类型" message:nil cancel:@"取消" actions:@[@"FWCacheMemory", @"FWCacheUserDefaults", @"FWCacheKeychain", @"FWCacheFile", @"FWCacheSqlite"] actionBlock:^(NSInteger index) {
            FWStrongifySelf();
            
            if (index == 0) {
                self.cache = [FWCacheManager managerWithType:FWCacheTypeMemory];
            } else if (index == 1) {
                self.cache = [FWCacheManager managerWithType:FWCacheTypeUserDefaults];
            } else if (index == 2) {
                self.cache = [FWCacheManager managerWithType:FWCacheTypeKeychain];
            } else if (index == 3) {
                self.cache = [FWCacheManager managerWithType:FWCacheTypeFile];
            } else if (index == 4) {
                self.cache = [FWCacheManager managerWithType:FWCacheTypeSqlite];
            }
            
            [self refreshCache];
        }];
    }];
}

- (void)setupSubviews
{
    UILabel *cacheLabel = [UILabel new];
    self.cacheLabel = cacheLabel;
    cacheLabel.numberOfLines = 0;
    cacheLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:cacheLabel]; {
        [cacheLabel fw_pinHorizontalToSuperviewWithInset:10];
        [cacheLabel fw_pinEdgeToSafeArea:NSLayoutAttributeTop withInset:10];
    }
    
    UIButton *refreshButton = [AppTheme largeButton];
    [refreshButton setTitle:@"读取缓存" forState:UIControlStateNormal];
    FWWeakifySelf();
    [refreshButton fw_addTouchBlock:^(id sender) {
        FWStrongifySelf();
        
        [self refreshCache];
    }];
    [self.view addSubview:refreshButton]; {
        [refreshButton fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:cacheLabel withOffset:10];
        [refreshButton fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    }
    
    UIButton *cacheButton = [AppTheme largeButton];
    [cacheButton setTitle:@"写入缓存" forState:UIControlStateNormal];
    [cacheButton fw_addTouchBlock:^(id sender) {
        FWStrongifySelf();
        
        [self.cache setObject:[NSUUID UUID].UUIDString forKey:TestCacheKey];
        [self refreshCache];
    }];
    [self.view addSubview:cacheButton]; {
        [cacheButton fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:refreshButton withOffset:10];
        [cacheButton fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    }
    
    UIButton *expireButton = [AppTheme largeButton];
    [expireButton setTitle:@"写入缓存(10s)" forState:UIControlStateNormal];
    [expireButton fw_addTouchBlock:^(id sender) {
        FWStrongifySelf();
        
        [self.cache setObject:[NSUUID UUID].UUIDString forKey:TestCacheKey withExpire:10];
        [self refreshCache];
    }];
    [self.view addSubview:expireButton]; {
        [expireButton fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:cacheButton withOffset:10];
        [expireButton fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    }
    
    UIButton *deleteButton = [AppTheme largeButton];
    [deleteButton setTitle:@"删除缓存" forState:UIControlStateNormal];
    [deleteButton fw_addTouchBlock:^(id sender) {
        FWStrongifySelf();
        
        [self.cache removeObjectForKey:TestCacheKey];
        [self refreshCache];
    }];
    [self.view addSubview:deleteButton]; {
        [deleteButton fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:expireButton withOffset:10];
        [deleteButton fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    }
    
    UIButton *clearButton = [AppTheme largeButton];
    [clearButton setTitle:@"清空缓存" forState:UIControlStateNormal];
    [clearButton fw_addTouchBlock:^(id sender) {
        FWStrongifySelf();
        
        [self.cache removeAllObjects];
        [self refreshCache];
    }];
    [self.view addSubview:clearButton]; {
        [clearButton fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:deleteButton withOffset:10];
        [clearButton fw_alignAxisToSuperview:NSLayoutAttributeCenterX];
    }
}

- (void)setupLayout
{
    self.cache = [FWCacheMemory sharedInstance];
    [self refreshCache];
}

- (void)refreshCache
{
    NSMutableString *statusStr = [[NSMutableString alloc] init];
    [statusStr appendString:NSStringFromClass([self.cache class])];
    [statusStr appendString:@"\n"];
    NSString *cacheStr = [self.cache objectForKey:TestCacheKey];
    if ([cacheStr fw_isNotEmpty]) {
        [statusStr appendString:cacheStr];
    } else {
        [statusStr appendString:@"缓存不存在"];
    }
    [statusStr appendString:@"\n"];
    NSNumber *expireNum = [self.cache objectForKey:TestExpireKey];
    if ([expireNum fw_isNotEmpty]) {
        [statusStr appendString:[NSString stringWithFormat:@"%.1fs有效", [expireNum doubleValue] - [[NSDate date] timeIntervalSince1970]]];
    } else {
        [statusStr appendString:cacheStr.fw_isNotEmpty ? @"永久有效" : @"缓存无效"];
    }
    self.cacheLabel.text = statusStr;
}

@end

File: Example/FWFramework/Classes/Module/Test/Service/TestCacheController.swift
//
//  TestCacheController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/1.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestCacheController: UIViewController {
    
    private var cache: CacheProtocol?
    
    private static var testCacheKey = "testCacheKey"
    private static var testExpireKey = "testCacheKey.__EXPIRE__"
    
    private lazy var cacheLabel: UILabel = {
        let result = UILabel()
        result.numberOfLines = 0
        result.textAlignment = .center
        return result
    }()
    
    private lazy var refreshButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("读取缓存", for: .normal)
        result.app.addTouch { [weak self] _ in
            self?.refreshCache()
        }
        return result
    }()
    
    private lazy var cacheButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("写入缓存", for: .normal)
        result.app.addTouch { [weak self] _ in
            self?.cache?.setObject(UUID().uuidString, forKey: TestCacheController.testCacheKey)
            self?.refreshCache()
        }
        return result
    }()
    
    private lazy var expireButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("写入缓存(10s)", for: .normal)
        result.app.addTouch { [weak self] _ in
            self?.cache?.setObject(UUID().uuidString, forKey: TestCacheController.testCacheKey, withExpire: 10)
            self?.refreshCache()
        }
        return result
    }()
    
    private lazy var deleteButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("删除缓存", for: .normal)
        result.app.addTouch { [weak self] _ in
            self?.cache?.removeObject(forKey: TestCacheController.testCacheKey)
            self?.refreshCache()
        }
        return result
    }()
    
    private lazy var clearButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("清空缓存", for: .normal)
        result.app.addTouch { [weak self] _ in
            self?.cache?.removeAllObjects()
            self?.refreshCache()
        }
        return result
    }()
    
}

extension TestCacheController: ViewControllerProtocol {
    
    func setupNavbar() {
        app.setRightBarItem(UIBarButtonItem.SystemItem.action) { [weak self] _ in
            self?.app.showSheet(title: "选择缓存类型", message: nil, actions: ["CacheMemory", "CacheUserDefaults", "CacheKeychain", "CacheFile", "CacheSqlite"], actionBlock: { index in
                if index == 0 {
                    self?.cache = CacheManager.manager(type: .memory)
                } else if index == 1 {
                    self?.cache = CacheManager.manager(type: .userDefaults)
                } else if index == 2 {
                    self?.cache = CacheManager.manager(type: .keychain)
                } else if index == 3 {
                    self?.cache = CacheManager.manager(type: .file)
                } else if index == 4 {
                    self?.cache = CacheManager.manager(type: .sqlite)
                }
                self?.refreshCache()
            })
        }
    }
    
    func setupSubviews() {
        view.addSubview(cacheLabel)
        view.addSubview(refreshButton)
        view.addSubview(cacheButton)
        view.addSubview(expireButton)
        view.addSubview(deleteButton)
        view.addSubview(clearButton)
    }
    
    func setupLayout() {
        cacheLabel.app.layoutChain
            .horizontal(10)
            .top(toSafeArea: 10)
        
        refreshButton.app.layoutChain
            .top(toViewBottom: cacheLabel, offset: 10)
            .centerX()
        
        cacheButton.app.layoutChain
            .top(toViewBottom: refreshButton, offset: 10)
            .centerX()
        
        expireButton.app.layoutChain
            .top(toViewBottom: cacheButton, offset: 10)
            .centerX()
        
        deleteButton.app.layoutChain
            .top(toViewBottom: expireButton, offset: 10)
            .centerX()
        
        clearButton.app.layoutChain
            .top(toViewBottom: deleteButton, offset: 10)
            .centerX()
        
        setupCache()
    }
    
}

extension TestCacheController {
    
    func setupCache() {
        cache = CacheMemory.shared
        refreshCache()
    }
    
    func refreshCache() {
        var statusStr = ""
        if let cacheClass = cache as? NSObject {
            statusStr += NSStringFromClass(type(of: cacheClass))
            statusStr += "\n"
        }
        
        var hasCache = false
        if let cacheStr = cache?.object(forKey: TestCacheController.testCacheKey) as? String, !cacheStr.isEmpty {
            statusStr += cacheStr
            hasCache = true
        } else {
            statusStr += "缓存不存在"
            hasCache = false
        }
        statusStr += "\n"
        
        if let expireNum = cache?.object(forKey: TestCacheController.testExpireKey) as? NSNumber {
            statusStr += String(format: "%.1fs有效", expireNum.doubleValue - Date().timeIntervalSince1970)
        } else {
            statusStr += hasCache ? "永久有效" : "缓存无效"
        }
        cacheLabel.text = statusStr
    }
    
}
